Look up Certificate ARNs that match ingress'

cv opened the pull request:
Hi! 👋 

I'm attempting a setup in which users of my k8s cluster get some subdomains to play with, like `<app>.<user>.dev.example.com`. I'd like to ensure all of their public-facing pods have an HTTP redirect to HTTPS, and that the certificate for their domain (say, `*.cv.dev.example.com`) is stored and managed by ACM (which is done elsewhere, in a Terraform script).

In order to minimize confusion, I'd like to use the smallest possible number of annotations, so that most apps will just work without a lot of yaml and helm template tweaking.

To do that, I'd like to have aws-alb-ingress-controller look that certificate up for me, so that I don't have to leak the abstraction by asking my users to know their certificate ARN. That way, if there is a certificate that matches a  `spec.tls.hosts[]` entry, I don't need to add the `certificate-arn` annotation to the ingresses.

In other words, if I have successfully issued a cert `arn:aws:acm:xxx:yyy:certificate/zzz` for the `*.cv.dev.example.com` domain, the following ingress...

```
apiVersion: extensions/v1beta1
kind: Ingress
metadata:
  annotations:
    alb.ingress.kubernetes.io/scheme: internet-facing
    alb.ingress.kubernetes.io/target-type: ip
    alb.ingress.kubernetes.io/listen-ports: '[{"HTTPS": 443}]'
    kubernetes.io/ingress.class: alb
  name: echoserver
  namespace: echoserver
spec:
  tls:
  - hosts:
    - echoserver.cv.dev.example.com
  rules:
  - host: echoserver.cv.dev.example.com
    http:
      paths:
      - backend:
          serviceName: echoserver
          servicePort: 80
        path: /
```

should:

- Set up a HTTP->HTTPS redirect (out of scope for this PR, believe I can do that with actions already)
- Set up a HTTPS listener with the `arn:aws:acm:xxx:yyy:certificate/zzz` cert

The implementation thus far is quite naive and almost pseudo-code, but should serve as a starting point to gather more feedback.

WDYT?

bigkraig commented:
Sounds good to me. Auto-discovery of certificates has come up in discussions in the past. 👍

cv commented:
@bigkraig excellent. Lemme add a couple of tests to the `getCertificate` func just to make sure I'm not doing anything stupid, then.

bigkraig commented:
May be worth a quick peek at https://github.com/kubernetes-sigs/aws-alb-ingress-controller/blob/master/internal/aws/albrgt/dummy.go to see how some of the mocked stuff has been done lately. We were talking about renaming it from dummy to mock, but the pattern is good for leveraging the mock client for tests in other packages.

cv commented:
Ready for review! 

Here are some outstanding questions:

#### Should we be concerned with the performance of `acm.ListCertificates` at this point?

AFAICT, this is a relatively low-pressure area of the code, so if the call takes long to return we'll just have to sit around and wait a little longer before our listener is created. Some installs can end up with quite a few certificates, but I don't have a scenario where thousands of certificates being returned would be realistic.

#### Should we be concerned with retry logic?

I don't know how often calls to `acm.ListCertificates` fail, and whether retrying is a good strategy in this specific code path. Might as well just fail the whole thing and let alb-ingress try the entire DesiredListener again later.

bigkraig commented:
https://github.com/kubernetes-sigs/aws-alb-ingress-controller/blob/master/internal/aws/albelbv2/dummy.go is also a good example of a more complicated client. 

Then see how the dummy client is set up in https://github.com/kubernetes-sigs/aws-alb-ingress-controller/blob/master/internal/alb/ls/listener_test.go#L40, it means you dont need to pass any parameters into the func.

cv commented:
@bigkraig I like those! Will refactor in a few minutes...

cv commented:
…done. AFAICT, this is ready to go, unless we want to do something about the [concerns mentioned above](https://github.com/kubernetes-sigs/aws-alb-ingress-controller/pull/600#issuecomment-420385748) before this hits master.

rifelpet commented on internal/alb/ls/listener.go:
```diff
@@ -333,3 +338,62 @@ func (l *Listener) DefaultActionArn() *string {
+	}
+
+	logger.Debugf("New desired listener wants HTTPS, but hasn't provided an certificate-arn annotation")
+	certificates, err := albacm.ACMsvc.ListCertificates(&acm.ListCertificatesInput{
```
Can we use the [paginated ListCertificates](https://docs.aws.amazon.com/sdk-for-go/api/service/acm/#ACM.ListCertificatesPages) instead?

rifelpet commented:
We have hundreds of ACM certs in the same region as our kubernetes clusters, I'm willing to test this branch out once the ListCertificates call is paginated and see how much of a performance impact it has.

cv commented:
Thanks for the offer, @rifelpet! I have updated the code to use `ListCertificatesPages`, can you please take it for a spin and see what you find? Cheers!

bigkraig commented:
hey @cv check out the `request.Pagination` stuff like in https://github.com/kubernetes-sigs/aws-alb-ingress-controller/blob/master/internal/aws/albrgt/rgt.go#L119

cv commented:
@bigkraig I did, but the `Pagination` struct docs state:

> A Pagination provides paginating of SDK API operations which are paginatable.
> Generally you should not use this type directly, but use the "Pages" API operations method to automatically perform pagination for you. Such as, "S3.ListObjectsPages", and "S3.ListObjectsPagesWithContext" methods.

…so I opted for using `ListCertificatesPages` as instructed.

cv commented:
Thinking about it some more, we can probably simplify the code for `GetClusterResources` a bit by using `GetResourcesPages`. I'll open another PR for that.

rifelpet commented:
I ran this branch and created a new Ingress listening on https and no certificate-arn set and got this crash:
```
I0912 16:51:05.585999       1 loadbalancer.go:265] test/acm: Start ELBV2 creation.
I0912 16:51:06.599420       1 loadbalancer.go:272] test/acm: Completed ELBV2 creation. Name: 7e18bd7e-test-acm-c179 | ARN: arn:aws:elasticloadbalancing:us-east-1:000000000000:loadbalancer/app/7e18bd7e-test-acm-c179/6eeefaa819d4d760
I0912 16:51:06.599442       1 targetgroup.go:225] test/acm: Start TargetGroup creation.
I0912 16:51:06.599869       1 event.go:221] Event(v1.ObjectReference{Kind:"Ingress", Namespace:"test", Name:"acm", UID:"09ebb0bb-b6ac-11e8-9153-121a570fedec", APIVersion:"extensions/v1beta1", ResourceVersion:"39458817", FieldPath:""}): type: 'Normal' reason: 'CREATE' 7e18bd7e-test-acm-c179 created
I0912 16:51:07.022923       1 targetgroup.go:230] test/acm: Succeeded TargetGroup creation. ARN: arn:aws:elasticloadbalancing:us-east-1:000000000000:targetgroup/7e18bd7e-3deeebaf6cd47c93f83/68a12237e6b8bc4e | Name: 7e18bd7e-3deeebaf6cd47c93f83.
I0912 16:51:07.022946       1 targetgroup.go:225] test/acm: Start TargetGroup creation.
I0912 16:51:07.023544       1 event.go:221] Event(v1.ObjectReference{Kind:"Ingress", Namespace:"test", Name:"acm", UID:"09ebb0bb-b6ac-11e8-9153-121a570fedec", APIVersion:"extensions/v1beta1", ResourceVersion:"39458817", FieldPath:""}): type: 'Normal' reason: 'CREATE' 7e18bd7e-3deeebaf6cd47c93f83 target group created
I0912 16:51:07.390520       1 targetgroup.go:230] test/acm: Succeeded TargetGroup creation. ARN: arn:aws:elasticloadbalancing:us-east-1:000000000000:targetgroup/7e18bd7e-3deeebaf6cd47c93f83/68a12237e6b8bc4e | Name: 7e18bd7e-3deeebaf6cd47c93f83.
I0912 16:51:07.390554       1 listener.go:202] test/acm: Start Listener creation.
I0912 16:51:07.391225       1 event.go:221] Event(v1.ObjectReference{Kind:"Ingress", Namespace:"test", Name:"acm", UID:"09ebb0bb-b6ac-11e8-9153-121a570fedec", APIVersion:"extensions/v1beta1", ResourceVersion:"39458817", FieldPath:""}): type: 'Normal' reason: 'CREATE' 7e18bd7e-3deeebaf6cd47c93f83 target group created
I0912 16:51:07.425838       1 listener.go:207] test/acm: Completed Listener creation. ARN: arn:aws:elasticloadbalancing:us-east-1:000000000000:listener/app/7e18bd7e-test-acm-c179/6eeefaa819d4d760/2d4a353b25baf229 | Port: 80 | Proto: HTTP.
I0912 16:51:07.425858       1 rule.go:143] test/acm: Start Rule creation.
I0912 16:51:07.426369       1 event.go:221] Event(v1.ObjectReference{Kind:"Ingress", Namespace:"test", Name:"acm", UID:"09ebb0bb-b6ac-11e8-9153-121a570fedec", APIVersion:"extensions/v1beta1", ResourceVersion:"39458817", FieldPath:""}): type: 'Normal' reason: 'CREATE' 80 listener created
I0912 16:51:07.452313       1 rule.go:148] test/acm: Completed Rule creation. Rule Priority: "1" | Condition: [{    Field: "host-header",    Values: ["testdomain.com"]  },{    Field: "path-pattern",    Values: ["/*"]  }]
I0912 16:51:07.452336       1 listener.go:202] test/acm: Start Listener creation.
I0912 16:51:07.452811       1 event.go:221] Event(v1.ObjectReference{Kind:"Ingress", Namespace:"test", Name:"acm", UID:"09ebb0bb-b6ac-11e8-9153-121a570fedec", APIVersion:"extensions/v1beta1", ResourceVersion:"39458817", FieldPath:""}): type: 'Normal' reason: 'CREATE' 1 rule created
F0912 16:51:07.466351       1 albingresses.go:167] ERROR: Work Unit failed due to a recoverable error: 'runtime error: invalid memory address or nil pointer dereference'
, Stack Trace:
 goroutine 869 [running]:
github.com/kubernetes-sigs/aws-alb-ingress-controller/vendor/gopkg.in/go-playground/pool%2ev3.(*limitedPool).newWorker.func1.1(0xc004a97f18, 0xc0041b7f00)
	/Users/$user/src/github.com/kubernetes-sigs/aws-alb-ingress-controller/vendor/gopkg.in/go-playground/pool.v3/limited_pool.go:60 +0xbc
panic(0x1688d60, 0x2c2bd90)
	/usr/local/Cellar/go/1.11/libexec/src/runtime/panic.go:513 +0x1b9
github.com/kubernetes-sigs/aws-alb-ingress-controller/internal/alb/ls.(*Listener).DefaultActionArn(...)
	/Users/$user/src/github.com/kubernetes-sigs/aws-alb-ingress-controller/internal/alb/ls/listener.go:336
github.com/kubernetes-sigs/aws-alb-ingress-controller/internal/alb/lb.(*LoadBalancer).Reconcile(0xc00430cdc0, 0xc004a97e30, 0xc004112d60, 0x10, 0x10)
	/Users/$user/src/github.com/kubernetes-sigs/aws-alb-ingress-controller/internal/alb/lb/loadbalancer.go:314 +0x405
github.com/kubernetes-sigs/aws-alb-ingress-controller/internal/albingress.(*ALBIngress).Reconcile(0xc004bc4a80, 0xc0044096a8, 0x0, 0x0)
	/Users/$user/src/github.com/kubernetes-sigs/aws-alb-ingress-controller/internal/albingress/albingress.go:281 +0x12e
github.com/kubernetes-sigs/aws-alb-ingress-controller/internal/albingress.ALBIngresses.Reconcile.func1.1(0x1ada1e0, 0xc00437ca20, 0x2, 0x0, 0x16ae801, 0xc000042060)
	/Users/$user/src/github.com/kubernetes-sigs/aws-alb-ingress-controller/internal/albingress/albingresses.go:154 +0xce
github.com/kubernetes-sigs/aws-alb-ingress-controller/vendor/gopkg.in/go-playground/pool%2ev3.(*limitedPool).newWorker.func1(0xc00437c8a0, 0xc004154e40, 0xc0041b7f00)
	/Users/$user/src/github.com/kubernetes-sigs/aws-alb-ingress-controller/vendor/gopkg.in/go-playground/pool.v3/limited_pool.go:89 +0x196
created by github.com/kubernetes-sigs/aws-alb-ingress-controller/vendor/gopkg.in/go-playground/pool%2ev3.(*limitedPool).newWorker
	/Users/$user/src/github.com/kubernetes-sigs/aws-alb-ingress-controller/vendor/gopkg.in/go-playground/pool.v3/limited_pool.go:52 +0x53

goroutine 1 [select]:
github.com/kubernetes-sigs/aws-alb-ingress-controller/internal/ingress/controller.(*ALBController).Start(0xc00041a2c0)
	/Users/$user/src/github.com/kubernetes-sigs/aws-alb-ingress-controller/internal/ingress/controller/alb.go:193 +0x3ff
main.main()
	/Users/$user/src/github.com/kubernetes-sigs/aws-alb-ingress-controller/cmd/main.go:117 +0x4fd

goroutine 5 [chan receive]:
github.com/kubernetes-sigs/aws-alb-ingress-controller/vendor/github.com/golang/glog.(*loggingT).flushDaemon(0x2c429a0)
	/Users/$user/src/github.com/kubernetes-sigs/aws-alb-ingress-controller/vendor/github.com/golang/glog/glog.go:882 +0x8b
created by github.com/kubernetes-sigs/aws-alb-ingress-controller/vendor/github.com/golang/glog.init.0
	/Users/$user/src/github.com/kubernetes-sigs/aws-alb-ingress-controller/vendor/github.com/golang/glog/glog.go:410 +0x203

goroutine 8 [syscall, 1 minutes]:
os/signal.signal_recv(0x0)
	/usr/local/Cellar/go/1.11/libexec/src/runtime/sigqueue.go:139 +0x9c
os/signal.loop()
	/usr/local/Cellar/go/1.11/libexec/src/os/signal/signal_unix.go:23 +0x22
created by os/signal.init.0
	/usr/local/Cellar/go/1.11/libexec/src/os/signal/signal_unix.go:29 +0x41

goroutine 48 [chan receive, 1 minutes]:
github.com/kubernetes-sigs/aws-alb-ingress-controller/vendor/k8s.io/client-go/tools/cache.(*sharedProcessor).run(0xc0001795e0, 0xc000444b40)
	/Users/$user/src/github.com/kubernetes-sigs/aws-alb-ingress-controller/vendor/k8s.io/client-go/tools/cache/shared_informer.go:425 +0x46
github.com/kubernetes-sigs/aws-alb-ingress-controller/vendor/k8s.io/client-go/tools/cache.(*sharedProcessor).run-fm(0xc000444b40)
	/Users/$user/src/github.com/kubernetes-sigs/aws-alb-ingress-controller/vendor/k8s.io/client-go/tools/cache/shared_informer.go:220 +0x34
github.com/kubernetes-sigs/aws-alb-ingress-controller/vendor/k8s.io/apimachinery/pkg/util/wait.(*Group).StartWithChannel.func1()
	/Users/$user/src/github.com/kubernetes-sigs/aws-alb-ingress-controller/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:54 +0x2e
github.com/kubernetes-sigs/aws-alb-ingress-controller/vendor/k8s.io/apimachinery/pkg/util/wait.(*Group).Start.func1(0xc00046a6b0, 0xc000422c00)
	/Users/$user/src/github.com/kubernetes-sigs/aws-alb-ingress-controller/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:71 +0x4f
created by github.com/kubernetes-sigs/aws-alb-ingress-controller/vendor/k8s.io/apimachinery/pkg/util/wait.(*Group).Start
	/Users/$user/src/github.com/kubernetes-sigs/aws-alb-ingress-controller/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:69 +0x62

goroutine 24 [select]:
github.com/kubernetes-sigs/aws-alb-ingress-controller/vendor/k8s.io/client-go/util/workqueue.(*delayingType).waitingLoop(0xc00009efa0)
	/Users/$user/src/github.com/kubernetes-sigs/aws-alb-ingress-controller/vendor/k8s.io/client-go/util/workqueue/delaying_queue.go:206 +0x34f
created by github.com/kubernetes-sigs/aws-alb-ingress-controller/vendor/k8s.io/client-go/util/workqueue.newDelayingQueue
	/Users/$user/src/github.com/kubernetes-sigs/aws-alb-ingress-controller/vendor/k8s.io/client-go/util/workqueue/delaying_queue.go:54 +0x1ba

goroutine 25 [chan receive]:
github.com/kubernetes-sigs/aws-alb-ingress-controller/vendor/k8s.io/apimachinery/pkg/watch.(*Broadcaster).loop(0xc000038c40)
	/Users/$user/src/github.com/kubernetes-sigs/aws-alb-ingress-controller/vendor/k8s.io/apimachinery/pkg/watch/mux.go:207 +0x66
created by github.com/kubernetes-sigs/aws-alb-ingress-controller/vendor/k8s.io/apimachinery/pkg/watch.NewBroadcaster
	/Users/$user/src/github.com/kubernetes-sigs/aws-alb-ingress-controller/vendor/k8s.io/apimachinery/pkg/watch/mux.go:75 +0xcc

goroutine 21 [IO wait]:
internal/poll.runtime_pollWait(0x7f833bf0af00, 0x72, 0xc00009d870)
	/usr/local/Cellar/go/1.11/libexec/src/runtime/netpoll.go:173 +0x66
internal/poll.(*pollDesc).wait(0xc00044b498, 0x72, 0xffffffffffffff00, 0x1abadc0, 0x2be34f0)
	/usr/local/Cellar/go/1.11/libexec/src/internal/poll/fd_poll_runtime.go:85 +0x9a
internal/poll.(*pollDesc).waitRead(0xc00044b498, 0xc0004fe000, 0x8000, 0x8000)
	/usr/local/Cellar/go/1.11/libexec/src/internal/poll/fd_poll_runtime.go:90 +0x3d
internal/poll.(*FD).Read(0xc00044b480, 0xc0004fe000, 0x8000, 0x8000, 0x0, 0x0, 0x0)
	/usr/local/Cellar/go/1.11/libexec/src/internal/poll/fd_unix.go:169 +0x179
net.(*netFD).Read(0xc00044b480, 0xc0004fe000, 0x8000, 0x8000, 0x408a0b, 0xc00002e000, 0x16ae960)
	/usr/local/Cellar/go/1.11/libexec/src/net/fd_unix.go:202 +0x4f
net.(*conn).Read(0xc00048e000, 0xc0004fe000, 0x8000, 0x8000, 0x0, 0x0, 0x0)
	/usr/local/Cellar/go/1.11/libexec/src/net/net.go:177 +0x68
crypto/tls.(*block).readFromUntil(0xc0000845d0, 0x1
```


```yaml
apiVersion: extensions/v1beta1
kind: Ingress
metadata:
  name: acm
  namespace: test
  annotations:
    kubernetes.io/ingress.class: alb
    alb.ingress.kubernetes.io/scheme: internal
    alb.ingress.kubernetes.io/subnets: xxx,yyy,zzz
    alb.ingress.kubernetes.io/security-groups: sg-12345678
    alb.ingress.kubernetes.io/tags: k8s=true
    alb.ingress.kubernetes.io/listen-ports: '[{"HTTP":80,"HTTPS":443}]'
    alb.ingress.kubernetes.io/attributes: deletion_protection.enabled=true,access_logs.s3.enabled=true,access_logs.s3.bucket=abcdefg
spec:
  rules:
  - host: testdomain.com
    http:
      paths:
      - path: /*
        backend:
          serviceName: test
          servicePort: 80
```

cv commented:
@rifelpet thanks for trying it out! 

This looks related to #594; I've added some debugging code to that area of the code and made it so that a panic is no longer possible. Should make it easier to figure out what's broken. Could you run it against the new build, please?

rifelpet commented:
Hmm I'm now getting this:
```
I0912 19:10:21.555764       1 loadbalancer.go:265] test/acm: Start ELBV2 creation.
I0912 19:10:22.506429       1 loadbalancer.go:272] test/acm: Completed ELBV2 creation. Name: 7e18bd7e-test-testacm-c179 | ARN: arn:aws:elasticloadbalancing:us-east-1:000000000000:loadbalancer/app/7e18bd7e-test-testacm-c179/24b743ee214c4d4e
I0912 19:10:22.506450       1 targetgroup.go:225] test/acm: Start TargetGroup creation.
I0912 19:10:22.506852       1 event.go:221] Event(v1.ObjectReference{Kind:"Ingress", Namespace"test", Nam"test", UID:"7c98fefa-b6bf-11e8-beb0-0e72888438a0", APIVersion:"extensions/v1beta1", ResourceVersion:"39477968", FieldPath:""}): type: 'Normal' reason: 'CREATE' 7e18bd7e-test-testacm-c179 created
I0912 19:10:22.971425       1 targetgroup.go:230] test/acm: Succeeded TargetGroup creation. ARN: arn:aws:elasticloadbalancing:us-east-1:000000000000:targetgroup/7e18bd7e-3deeebaf6cd47c93f83/ff4b5ab29b5e9014 | Name: 7e18bd7e-3deeebaf6cd47c93f83.
I0912 19:10:22.971474       1 targetgroup.go:225] test/acm: Start TargetGroup creation.
I0912 19:10:22.971999       1 event.go:221] Event(v1.ObjectReference{Kind:"Ingress", Namespace"test", Nam"test", UID:"7c98fefa-b6bf-11e8-beb0-0e72888438a0", APIVersion:"extensions/v1beta1", ResourceVersion:"39477968", FieldPath:""}): type: 'Normal' reason: 'CREATE' 7e18bd7e-3deeebaf6cd47c93f83 target group created
I0912 19:10:23.332676       1 targetgroup.go:230] test/acm: Succeeded TargetGroup creation. ARN: arn:aws:elasticloadbalancing:us-east-1:000000000000:targetgroup/7e18bd7e-3deeebaf6cd47c93f83/ff4b5ab29b5e9014 | Name: 7e18bd7e-3deeebaf6cd47c93f83.
I0912 19:10:23.332702       1 listener.go:202] test/acm: Start Listener creation.
I0912 19:10:23.333092       1 event.go:221] Event(v1.ObjectReference{Kind:"Ingress", Namespace"test", Nam"test", UID:"7c98fefa-b6bf-11e8-beb0-0e72888438a0", APIVersion:"extensions/v1beta1", ResourceVersion:"39477968", FieldPath:""}): type: 'Normal' reason: 'CREATE' 7e18bd7e-3deeebaf6cd47c93f83 target group created
E0912 19:10:23.342788       1 albingress.go:290] test/acm: Failed to reconcile state on this ingress
E0912 19:10:23.342808       1 albingress.go:292] test/acm:  - Failed Listener creation: ValidationError: A certificate must be specified for HTTPS listeners
E0912 19:10:23.342813       1 albingress.go:292] test/acm: 	status code: 400, request id: 7fc3fe94-b6bf-11e8-a115-c3a644444f0e.
E0912 19:10:23.342823       1 albingress.go:295] test/acm: Will retry to reconcile in 497.675119ms
I0912 19:10:23.342993       1 event.go:221] Event(v1.ObjectReference{Kind:"Ingress", Namespace"test", Nam"test", UID:"7c98fefa-b6bf-11e8-beb0-0e72888438a0", APIVersion:"extensions/v1beta1", ResourceVersion:"39477968", FieldPath:""}): type: 'Warning' reason: 'ERROR' Error creating 443 listener: ValidationError: A certificate must be specified for HTTPS listeners
	status code: 400, request id: 7fc3fe94-b6bf-11e8-a115-c3a644444f0e
```

cv commented:
@rifelpet can you turn on debug level logging and try again, please? 

I'm not sure if there's an easier way, but adding `-v2` to the `spec.containers.args` list of the alb-ingress-controller Deployment yaml should do the trick.

cv commented:
Oh, and just checking: sure you have an *issued* certificate for `testdomain.com` in ACM?

rifelpet commented:
That domain I tested actually used a wildcard ACM cert. Its probably out of the scope of this PR to support falling back to wildcards if an explicit cert doesnt exist, but that would be nice to have down the road.

I did create one with a different domain that does have a matching issued certificate:

```
I0912 21:45:44.176422       1 store.go:206] Event(v1.ObjectReference{Kind:"Ingress", Namespace:"test", Name:"acm", UID:"3368df1a-b6d5-11e8-9153-121a570fedec", APIVersion:"extensions/v1beta1", ResourceVersion:"39499271", FieldPath:""}): type: 'Normal' reason: 'CREATE' Ingress test/acm
I0912 21:45:44.177234       1 listener.go:364] test/acm: New desired listener wants HTTPS, but hasn't provided an certificate-arn annotation
I0912 21:45:44.392474       1 listener.go:374] test/acm: 100 issued certificates in AWS ACM response page 0
I0912 21:45:44.464562       1 listener.go:374] test/acm: 100 issued certificates in AWS ACM response page 1
I0912 21:45:44.534208       1 listener.go:374] test/acm: 100 issued certificates in AWS ACM response page 2
I0912 21:45:44.602005       1 listener.go:374] test/acm: 100 issued certificates in AWS ACM response page 3
I0912 21:45:44.666502       1 listener.go:374] test/acm: 55 issued certificates in AWS ACM response page 4
I0912 21:45:44.668308       1 loadbalancer.go:265] test/acm: Start ELBV2 creation.
I0912 21:45:44.668607       1 listener.go:341] test/test-internal: DefaultAction: l.ls.current.DefaultActions is empty
I0912 21:45:44.668617       1 listener.go:341] test/test-internal: DefaultAction: l.ls.current.DefaultActions is empty
I0912 21:45:46.073273       1 loadbalancer.go:272] test/acm: Completed ELBV2 creation. Name: 7e18bd7e-test-acm-c179 | ARN: arn:aws:elasticloadbalancing:us-east-1:000000000000:loadbalancer/app/7e18bd7e-test-acm-c179/480f3803f050b21e
I0912 21:45:46.073300       1 targetgroup.go:225] test/acm: Start TargetGroup creation.
I0912 21:45:46.073574       1 event.go:221] Event(v1.ObjectReference{Kind:"Ingress", Namespace:"test", Name:"acm", UID:"3368df1a-b6d5-11e8-9153-121a570fedec", APIVersion:"extensions/v1beta1", ResourceVersion:"39499271", FieldPath:""}): type: 'Normal' reason: 'CREATE' 7e18bd7e-test-acm-c179 created
I0912 21:45:46.519882       1 targetgroup.go:230] test/acm: Succeeded TargetGroup creation. ARN: arn:aws:elasticloadbalancing:us-east-1:000000000000:targetgroup/7e18bd7e-3deeebaf6cd47c93f83/ec0bc37b2df64dab | Name: 7e18bd7e-3deeebaf6cd47c93f83.
I0912 21:45:46.519906       1 targetgroup.go:225] test/acm: Start TargetGroup creation.
I0912 21:45:46.520182       1 event.go:221] Event(v1.ObjectReference{Kind:"Ingress", Namespace:"test", Name:"acm", UID:"3368df1a-b6d5-11e8-9153-121a570fedec", APIVersion:"extensions/v1beta1", ResourceVersion:"39499271", FieldPath:""}): type: 'Normal' reason: 'CREATE' 7e18bd7e-3deeebaf6cd47c93f83 target group created
I0912 21:45:47.076730       1 targetgroup.go:230] test/acm: Succeeded TargetGroup creation. ARN: arn:aws:elasticloadbalancing:us-east-1:000000000000:targetgroup/7e18bd7e-3deeebaf6cd47c93f83/ec0bc37b2df64dab | Name: 7e18bd7e-3deeebaf6cd47c93f83.
I0912 21:45:47.076756       1 listener.go:202] test/acm: Start Listener creation.
I0912 21:45:47.077221       1 event.go:221] Event(v1.ObjectReference{Kind:"Ingress", Namespace:"test", Name:"acm", UID:"3368df1a-b6d5-11e8-9153-121a570fedec", APIVersion:"extensions/v1beta1", ResourceVersion:"39499271", FieldPath:""}): type: 'Normal' reason: 'CREATE' 7e18bd7e-3deeebaf6cd47c93f83 target group created
I0912 21:45:47.121185       1 listener.go:207] test/acm: Completed Listener creation. ARN: arn:aws:elasticloadbalancing:us-east-1:000000000000:listener/app/7e18bd7e-test-acm-c179/480f3803f050b21e/665482f9ddb09096 | Port: 80 | Proto: HTTP.
I0912 21:45:47.121212       1 rule.go:143] test/acm: Start Rule creation.
I0912 21:45:47.121874       1 event.go:221] Event(v1.ObjectReference{Kind:"Ingress", Namespace:"test", Name:"acm", UID:"3368df1a-b6d5-11e8-9153-121a570fedec", APIVersion:"extensions/v1beta1", ResourceVersion:"39499271", FieldPath:""}): type: 'Normal' reason: 'CREATE' 80 listener created
I0912 21:45:47.231072       1 rule.go:148] test/acm: Completed Rule creation. Rule Priority: "1" | Condition: [{    Field: "path-pattern",    Values: ["/*"]  },{    Field: "host-header",    Values: ["testdomain.com"]  }]
I0912 21:45:47.231096       1 listener.go:202] test/acm: Start Listener creation.
I0912 21:45:47.231367       1 event.go:221] Event(v1.ObjectReference{Kind:"Ingress", Namespace:"test", Name:"acm", UID:"3368df1a-b6d5-11e8-9153-121a570fedec", APIVersion:"extensions/v1beta1", ResourceVersion:"39499271", FieldPath:""}): type: 'Normal' reason: 'CREATE' 1 rule created
I0912 21:45:47.258472       1 listener.go:341] test/acm: DefaultAction: l.ls.current.DefaultActions is empty
I0912 21:45:47.258496       1 listener.go:337] test/acm: DefaultAction: l.ls.current is nil
E0912 21:45:47.258662       1 albingress.go:290] test/acm: Failed to reconcile state on this ingress
E0912 21:45:47.258745       1 albingress.go:292] test/acm:  - Failed Listener creation: ValidationError: A certificate must be specified for HTTPS listeners
E0912 21:45:47.258756       1 albingress.go:292] test/acm: 	status code: 400, request id: 353d9e9f-b6d5-11e8-96cd-ef0e7d978a63.
E0912 21:45:47.258767       1 albingress.go:295] test/acm: Will retry to reconcile in 503.246016ms
I0912 21:45:47.258823       1 event.go:221] Event(v1.ObjectReference{Kind:"Ingress", Namespace:"test", Name:"acm", UID:"3368df1a-b6d5-11e8-9153-121a570fedec", APIVersion:"extensions/v1beta1", ResourceVersion:"39499271", FieldPath:""}): type: 'Warning' reason: 'ERROR' Error creating 443 listener: ValidationError: A certificate must be specified for HTTPS listeners
	status code: 400, request id: 353d9e9f-b6d5-11e8-96cd-ef0e7d978a63
```

The good news is that the paginated calls appear to be pretty fast which was my initial concern.

cv commented:
Oh, this is much better already. It seems like you didn't include a `spec.tls.hosts` section in the ingress yaml, which is where I was looking for the host data.

Thinking about it though, I believe we should look for hosts *both* in the `spec.tls.hosts` block and in `spec.rules.host` instances. I've updated the code accordingly, if you'd like to take it for a spin.

rifelpet commented:
Looks much better, that did the trick!

```
I0912 22:05:43.400020       1 store.go:206] Event(v1.ObjectReference{Kind:"Ingress", Namespace:"test", Name:"acm", UID:"fe34c321-b6d7-11e8-8beb-0aded5f62134", APIVersion:"extensions/v1beta1", ResourceVersion:"39502022", FieldPath:""}): type: 'Normal' reason: 'CREATE' Ingress test/acm
I0912 22:05:44.609420       1 listener.go:364] test/acm: New desired listener wants HTTPS, but hasn't provided an certificate-arn annotation
I0912 22:05:44.881425       1 listener.go:374] test/acm: 455 issued certificates in AWS ACM response page 0
I0912 22:05:44.881449       1 listener.go:381] test/acm: Ignoring domain name '0xc001ce1020', doesn't match 'testdomain.com'
... repeat many times
I0912 22:05:44.891474       1 listener.go:381] test/acm: Ignoring domain name '0xc001cf99d0', doesn't match 'testdomain.com'
I0912 22:05:44.891515       1 listener.go:378] test/acm: Domain name '0xc001cf9a00', matches rule host 'testdomain.com', adding to Listener
I0912 22:05:44.891540       1 listener.go:381] test/acm: Ignoring domain name '0xc001cf9a30', doesn't match 'testdomain.com'
... repeat many times
I0912 22:05:44.895279       1 loadbalancer.go:265] test/acm: Start ELBV2 creation.
I0912 22:05:45.990866       1 loadbalancer.go:272] test/acm: Completed ELBV2 creation. Name: 7e18bd7e-test-acm-c179 | ARN: arn:aws:elasticloadbalancing:us-east-1:000000000000:loadbalancer/app/7e18bd7e-test-acm-c179/f480858d12506422
I0912 22:05:45.990891       1 targetgroup.go:225] test/acm: Start TargetGroup creation.
I0912 22:05:45.991671       1 event.go:221] Event(v1.ObjectReference{Kind:"Ingress", Namespace:"test", Name:"acm", UID:"fe34c321-b6d7-11e8-8beb-0aded5f62134", APIVersion:"extensions/v1beta1", ResourceVersion:"39502022", FieldPath:""}): type: 'Normal' reason: 'CREATE' 7e18bd7e-test-acm-c179 created
I0912 22:05:46.429545       1 targetgroup.go:230] test/acm: Succeeded TargetGroup creation. ARN: arn:aws:elasticloadbalancing:us-east-1:000000000000:targetgroup/7e18bd7e-3deeebaf6cd47c93f83/ec0bc37b2df64dab | Name: 7e18bd7e-3deeebaf6cd47c93f83.
I0912 22:05:46.429569       1 targetgroup.go:225] test/acm: Start TargetGroup creation.
I0912 22:05:46.430057       1 event.go:221] Event(v1.ObjectReference{Kind:"Ingress", Namespace:"test", Name:"acm", UID:"fe34c321-b6d7-11e8-8beb-0aded5f62134", APIVersion:"extensions/v1beta1", ResourceVersion:"39502022", FieldPath:""}): type: 'Normal' reason: 'CREATE' 7e18bd7e-3deeebaf6cd47c93f83 target group created
I0912 22:05:46.886114       1 targetgroup.go:230] test/acm: Succeeded TargetGroup creation. ARN: arn:aws:elasticloadbalancing:us-east-1:000000000000:targetgroup/7e18bd7e-3deeebaf6cd47c93f83/ec0bc37b2df64dab | Name: 7e18bd7e-3deeebaf6cd47c93f83.
I0912 22:05:46.886139       1 listener.go:202] test/acm: Start Listener creation.
I0912 22:05:46.886466       1 event.go:221] Event(v1.ObjectReference{Kind:"Ingress", Namespace:"test", Name:"acm", UID:"fe34c321-b6d7-11e8-8beb-0aded5f62134", APIVersion:"extensions/v1beta1", ResourceVersion:"39502022", FieldPath:""}): type: 'Normal' reason: 'CREATE' 7e18bd7e-3deeebaf6cd47c93f83 target group created
I0912 22:05:46.913823       1 listener.go:207] test/acm: Completed Listener creation. ARN: arn:aws:elasticloadbalancing:us-east-1:000000000000:listener/app/7e18bd7e-test-acm-c179/f480858d12506422/f8c778c13809da18 | Port: 80 | Proto: HTTP.
I0912 22:05:46.913845       1 rule.go:143] test/acm: Start Rule creation.
I0912 22:05:46.914514       1 event.go:221] Event(v1.ObjectReference{Kind:"Ingress", Namespace:"test", Name:"acm", UID:"fe34c321-b6d7-11e8-8beb-0aded5f62134", APIVersion:"extensions/v1beta1", ResourceVersion:"39502022", FieldPath:""}): type: 'Normal' reason: 'CREATE' 80 listener created
I0912 22:05:46.961036       1 rule.go:148] test/acm: Completed Rule creation. Rule Priority: "1" | Condition: [{    Field: "path-pattern",    Values: ["/*"]  },{    Field: "host-header",    Values: ["testdomain.com"]  }]
I0912 22:05:46.961089       1 listener.go:202] test/acm: Start Listener creation.
I0912 22:05:46.961657       1 event.go:221] Event(v1.ObjectReference{Kind:"Ingress", Namespace:"test", Name:"acm", UID:"fe34c321-b6d7-11e8-8beb-0aded5f62134", APIVersion:"extensions/v1beta1", ResourceVersion:"39502022", FieldPath:""}): type: 'Normal' reason: 'CREATE' 1 rule created
I0912 22:05:47.677705       1 listener.go:207] test/acm: Completed Listener creation. ARN: arn:aws:elasticloadbalancing:us-east-1:000000000000:listener/app/7e18bd7e-test-acm-c179/f480858d12506422/f77e6135ce555c00 | Port: 443 | Proto: HTTPS.
I0912 22:05:47.677737       1 rule.go:143] test/acm: Start Rule creation.
I0912 22:05:47.678186       1 event.go:221] Event(v1.ObjectReference{Kind:"Ingress", Namespace:"test", Name:"acm", UID:"fe34c321-b6d7-11e8-8beb-0aded5f62134", APIVersion:"extensions/v1beta1", ResourceVersion:"39502022", FieldPath:""}): type: 'Normal' reason: 'CREATE' 443 listener created
I0912 22:05:47.762692       1 rule.go:148] test/acm: Completed Rule creation. Rule Priority: "1" | Condition: [{    Field: "path-pattern",    Values: ["/*"]  },{    Field: "host-header",    Values: ["testdomain.com"]  }]
I0912 22:05:47.762716       1 listener.go:341] test/acm: DefaultAction: l.ls.current.DefaultActions is empty
I0912 22:05:47.762722       1 listener.go:341] test/acm: DefaultAction: l.ls.current.DefaultActions is empty
I0912 22:05:47.762835       1 event.go:221] Event(v1.ObjectReference{Kind:"Ingress", Namespace:"test", Name:"acm", UID:"fe34c321-b6d7-11e8-8beb-0aded5f62134", APIVersion:"extensions/v1beta1", ResourceVersion:"39502022", FieldPath:""}): type: 'Normal' reason: 'CREATE' 1 rule created
```

The response time for 500 per page still seems fine. In addition to supporting wildcards it would be helpful to support SANs, but according to the AWS API that requires calling DescribeCertificate on each cert so I'm not sure how feasible that would be. You could cache the results since certs are immutable but that would still be a lot of API calls up front.

Thanks for working on this!

cv commented:
@rifelpet not sure about SANs either, since it's likely that all those API requests in a short time span would get a big account like yours into some kind of trouble.

Wildcards should be supported, though. If you have a certificate for `*.example.com` and your ingress looks like this:

```
apiVersion: extensions/v1beta1
kind: Ingress
metadata:
  name: acm
  namespace: test
  annotations:
    kubernetes.io/ingress.class: alb
    alb.ingress.kubernetes.io/scheme: internal
    alb.ingress.kubernetes.io/subnets: xxx,yyy,zzz
    alb.ingress.kubernetes.io/security-groups: sg-12345678
    alb.ingress.kubernetes.io/tags: k8s=true
    alb.ingress.kubernetes.io/listen-ports: '[{"HTTP":80,"HTTPS":443}]'
    alb.ingress.kubernetes.io/attributes: deletion_protection.enabled=true,access_logs.s3.enabled=true,access_logs.s3.bucket=abcdefg
spec:
  rules:
  - host: foo.example.com
    http:
      paths:
      - path: /*
        backend:
          serviceName: test
          servicePort: 80
```

…it should pick it up just fine.

cv commented:
/lint

rifelpet commented on internal/alb/ls/listener.go:
```diff
@@ -328,8 +333,113 @@ func (l *Listener) GetRules() rs.Rules {
+	return certs, nil
+}
+
+func domainMatchesIngressTLSHost(domainName string, tlsHost string) bool {
```
Should `TLS` be removed from this function name given that we're now using both TLS hosts and Rules hosts?

cv replied:
Good call!

rifelpet commented:
Just tried the wildcard certificate with the latest changes in this branch and it seems to work fine now. Thanks for the assistance.

cv commented:
Oops. I had merged in some stuff that I shouldn't. Rebased and force-pushed. Sorry about making the comment thread look a bit weird now!

While at it, I've been trying to reproduce the issue @hrzbrg pointed out, and I just added a test case to make sure it's not in the `getCertificate` code. Looks to be happening further down the line as we try to create the actual load balancer.

cv commented:
Just saw this in the logs: 

```
Warning  ERROR  1m               aws-alb-ingress-controller  Error creating 443 listener: TooManyCertificates: Up to '1' certificate ARNs can be specified, but '2' were specified
           status code: 400, request id: 8d7b8fc7-c003-11e8-af7c-0dc2f94b60d2
```

I should've read more of the code :) `api.go` states:

```golang
type CreateListenerInput struct {
	_ struct{} `type:"structure"`

	// [HTTPS listeners] The default SSL server certificate. You must provide exactly
	// one default certificate. To create a certificate list, use AddListenerCertificates.
	Certificates []*Certificate `type:"list"`
...
```

Added a call to `AddListenerCertificates` when more than a single cert is found. Doing some manual testing to make sure AWS likes that change now.

hatemosphere commented:
@bigkraig could you please explain why this PR was closed?

bigkraig commented:
@hatemosphere we rebased the master branch which closed all of the existing PRs, you can see the impact it had if you look at the commits in here now. The best way to get this one back in order is to cherry-pick the commits into a branch that has been created off of the new master.

khyew commented:
@cv looks like your PR here got closed accidentally after a rebase of `master`. I'm still interested in your feature but the rebase also nuked GitHub's copy of your commits.

Would you mind cherry-picking them into a new branch and creating a new PR?

cv commented:
@khyew thanks for letting me know! I am no longer working on this, but will try to cherry-pick all the changes back in. Gimme a few minutes…

cv commented:
Well, that was overly enthusiastic of my part… I can't figure out how to cherry-pick the changes, or get a patch that applies at all. @khyew if you want to take a stab at it, all of the commits in the range from 54f733388859d45c24eebd72c8a08d9918040507 to 8ad75b3a73e1617a36d89b30e2f0ae83cab5b4a2 (in the original tree, that is) need to come back.

khyew commented:
@cv just the commits that were authored by you in that range, correct?

cv commented:
@khyew yup! I'm pretty sure I was the only one to commit to the branch.

khyew commented:
How did you test your PR? Were you able to deploy your own branch of the controller to a cluster? I found this guide btw: https://kubernetes-sigs.github.io/aws-alb-ingress-controller/BUILDING/

cv commented:
Exactly, built a new Docker image, pushed it to ECR and ran it on my EKS cluster from there.
